### ISSUE TITLE
Research: Stripe Link with deferred intent UPE

### ISSUE THREAD

**Issue by FangedParakeet** [Automattic/woocommerce-payments#6392]
### Description

Stripe Link must be supported within our deferred intent UPE implementation--for more information on the specifics of the shape of this implementation, please see #6391. 

However, before we are able to plan and proceed with this implementation, it would be very helpful for us to gather some more knowledge about the existing implementation and what will be required to port this feature over to the deferred intent UPE. Here are a few useful questions that we would like to find answers for.

- How does the existing implementation work?
- Which parts of the existing implementation are readily reusable?
- Which parts of the existing implementation will need to be refactored/extracted?
- Which parts of the new implementation will need to be create newly from scratch?

The result of this research can be a comment explaining findings on this issue, but please be ready to bring your findings with you to our squad meeting next week (June 2nd) to share this knowledge and discuss the ramifications with the rest of our squad as well.

### Additional context

#6391

**Comment by FangedParakeet** [comment]
# The basics

Stripe Link is an checkout method that--similar to WooPay--identifies users by their email address and authenticates with an SMS OTP. 

# The ropes

There are several different integration flows for Link offered by Stripe. I'm going to list them out and briefly describe them here, explaining why we are unfortunately restricted to using one specific flow that complies with our checkout.

- [Standard payment flow (Link Authentication Element)](https://stripe.com/docs/payments/link/accept-a-payment?platform=web&ui=elements&link-integration-type=at-payment#design-your-integration)
   - Email and billing fields replaced by Stripe elements. 
   - Conflicts with WC core, since we are not able to use our own inputs on the checkout page.
   - **Verdict: ❌**
- [Standard payment flow (without Link Authentication Element)](https://stripe.com/docs/payments/link/accept-a-payment?platform=web&ui=elements&link-integration-type=before-payment#design-your-integration)
   - Collect email address and manually pass it to payment element.
   - Email address can only be passed to element prior to mounting payment element.
   - In most cases, we will want to mount the payment element (on page load) and then collect email/billing information (prior to form submission).
   - **Verdict: ❌**
- [Link in the card element](https://stripe.com/docs/payments/link-card-element)
   - Link is added as an embedded button in the card payment element.
   - Clicking the button opens up authentication modal in new window overlay.
   - Only works with `card` payment element type, not the `payment` type used with the UPE. 
   - Requires us to use split-UPE to maintain both UPE and legacy gateways alongside each other...which is what we currently do, but is also what we're aiming to move away from.
   - **Verdict: ❌**
- [Link as a payment request button](https://stripe.com/docs/stripe-js/elements/payment-request-button?client=html#link-prb)
   - Link appears as a payment request button, like Apple/Google Pay.
   - Link will only appear when Apple/Google Pay are not available or customer has authenticated Chrome with Link.
   - This would restrict Link's ability to appear at checkout and we would have to include it within our Apple/Google Pay express checkout option.
   - **Verdict: ❌**
- [Link autofill modal](https://stripe.com/docs/payments/link/autofill-modal)
   - Use our own email input to trigger Link autofill modal. 
   - Successfully authenticated modal returns stored billing information, which we can use to prefill checkout fields.
   - Combines the `payment` element with our own email input and gives us some control over when and how the modal will appear.
   - **Verdict: ✅ It'll do.**
   
# The filling

So it appears that the autofill modal is still the only option available to us: now how exactly does it work?

The [Link Autofill modal](https://stripe.com/docs/payments/link/autofill-modal#reference) is a somewhat limited interface provided by Stripe that offers us three methods and two events to integrate our implementation.

The main function `launch` takes the email address as a parameter and is the main trigger for the experience. If the email address is identified as a registered user, the OTP modal will open; if not, nothing will happen and the optional sign-up checkbox will be added to the payment element instead. Note that this function does not return anything, so we will not be able to ascertain whether a user's email address has been recognised, until after they are authenticated.

There are two events that we can track. The `authenticated` event will fire after a user has successfully authenticated via OTP. The `autofill` event will fire after an authenticated user has chosen their desired payment/billing information and will return these credentials to allow us to prefill this information in the checkout form.

If the customer closes the autofill modal, the `show` function will reopen the model, for the customer to select their payment information again, after the user has been authenticated. Note that if a user closes the OTP modal _before_ they are authenticated, there is no way for us to reopen this modal. Finally, the `logout` function will end the experience for an authenticated user.

To tie everything together, we must pass the Stripe Elements object representing the payment element in question, as a parameter to the `stripe.linkAutofillModal()` function and this will return us a hydrated class linked to our payment element. We superimpose the Link logo onto the email field input and add a keyup listener that sends the value of the input to the `launch` function above. These both act as triggers to launch the experience. 

Finally, we add listeners and handlers for the `authenticated` and `autofill` events that fill in our checkout fields. 

All of this functionality is already implemented [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/stripe-link/index.js) and I don't believe this will require any modification. To add the Link integration, we just need to call the `enableStripeLinkPaymentMethod()` function, passing in the ID of the email input and directions of how to fill in billing/shipping information for the `autofill` event. 

We can find implementations of this function for the shortcode and blocks checkout [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/classic/upe-split.js#L244) and [here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/blocks/upe-split-fields.js#L126) respectively. I believe this is essentially all we will need to add for the new deferred intent UPE and since we are still keeping WooPay and Link separated, I don't think we should have any issues with compatibility. 

# The estimate

As long as there no issues on Stripe's end with using Link without an intent--which I have no reason to believe there would be, since the Link autofill constructor only requires the Stripe elements object, with no mention of intent nor client secret--I think this should be a very simple integration to accomplish, since we can reuse the existing implementation interface entirely and AFAIA it needs only to be called once in order to tie all the strings together.

As long as there are no unseen circumstances, I would measure this at **`2`-`3` points**. Stripe Link has decent [unit test coverage](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/stripe-link/test/index.test.js), though no E2E coverage AFAICT. That being said, E2E tests may be challenging to create, since I'm not sure that creating Link test accounts is particularly trivial and I'm not sure if we are able to parse elements from Stripe's iframed content. I would not consider this to be required for the completion of #6391 and, if desired, this can be included in a separate new issue.

**Comment by FangedParakeet** [comment]
Edit: I've converted this into a more extensive yet digestible post over here -> pbIwZe-1so-p2

**Comment by FangedParakeet** [comment]
Following my exploratory examinations and ensuing exchanges in pbIwZe-1so-p2, we have decided that Link's integration as an autofill modal remains the best available option from both a development and user experience perspective. This is very beneficial to us, because it means we can reuse the existing integration almost entirely. 

There is some uncertainty about whether Link in the payment element will allow us to defer intent creation and confirmation, similar to all other payment methods, so for this reason we have inflated an initially overtly optimistic estimate of effort to a still modest `3` points for #6391.
